packages/runtimes/js/src/helpers/browser/esm-js-loader-retry.js

```
@@ -1,42 +1,61 @@
+  const retryState = parcelRequire._retryState;
 
+  // The number of retries before rethrowing the error
   const maxRetries = 6;
 
-  for (let i = 1; i <= maxRetries; i++) {
-    // Wait for the user to go online before making a request
-    if (!globalThis.navigator.onLine) {
-      await new Promise((resolve) =>
-        globalThis.addEventListener('online', resolve, {once: true}),
-      );
-    }
-
-    let requestUrl = url;
-    if (i !== 0) {
-      // Date ensures the client hasn't previously cached a failed request
-      requestUrl = `${requestUrl}?retry=${i}:${Date.now()}`;
-    }
-
-    try {
-      // eslint-disable-next-line no-undef
-      return await __parcel__import__(requestUrl);
-    } catch (error) {
-      if (i === maxRetries) throw error;
-      // Dispatch event for reporting
-      window.dispatchEvent(
-        new CustomEvent('atlaspack:import_retry', {
-          detail: {
-            target: url,
-            attempt: i,
-          },
-        }),
-      );
-      const jitter = Math.round(Math.random() * 100);
-      const delay = Math.min(Math.pow(2, i), 8) * 1000;
-
-      await new Promise((resolve) => setTimeout(resolve, delay + jitter));
-    }
+  // Resolve the request URL from the bundle ID
+  const url = require('../bundle-manifest').resolve(id);
+
+  // Wait for the user to go online before making a request
+  if (!globalThis.navigator.onLine) {
+    await new Promise((resolve) =>
+      globalThis.addEventListener('online', resolve, {once: true}),
+    );
+  }
+
+  // If the import has not run or is not currently running
+  // then start the import retry task. Otherwise reuse the
+  // existing result or wait for the current task to complete
+  if (!retryState[url]) {
+    retryState[url] = (async () => {
+      // Try first request with normal import circuit
+      try {
+        // eslint-disable-next-line no-undef
+        return await __parcel__import__(url);
+      } catch {
+        /**/
```

Doesn't this mean it'll always retry as we ignore the error?

---

Ohh ok, I think I get what you're going for. How come you're not using the catch handler here? Reads a bit confusing.

---

This logic will only run once the first time `load` is called as the promise is cached in `retryState[url]` which is reused on subsequent calls to `load`

The flow goes:

- `retryState[url]` is immediately set to a `Promise<void>` defined as the import task
  - subsequent requests will get `retryState[url]` which will be the pending import
  
- The first import attempt uses the normal import logic
  - If the first import fails it begins retrying
    - if a retry attempt succeeds then `retryState[url]` resolves
    - if all retry attempts fail then ``retryState[url]` rejects

---

> How come you're not using the catch handler here? Reads a bit confusing.

Being a bit defensive about when retries start to occur. I wanted the original retry logic to exist as it did before before the retry circuit is kicked off.

I had it as part of the loop before which is fine too - just wanted to seperate it to be a little more explicit. Can refactor this later after I add some tests